### ISSUE TITLE
Support for multiple daqstreams

### ISSUE THREAD

**Issue by agamemnonc** [axopy/axopy#62]
In my understanding, it is not currently possible to support multiple daqs running simultaneously (e.g. EEG with EMG, or EMG with data glove etc.).

I have thought of a trivial solution but I think it would only work if the different sampling rates are perfect multiples of one another, so that the `samples_per_read` parameters are such that it takes exactly the same time to perform a `read()`  operation from the devices.

@ixjlyons Please let me know if this is something worth considering and, if so, I could open a PR and take it from there.

**Comment by ixjlyons** [comment]
Yeah, I couldn't really decide on an interface I liked to support multiple input devices. A recurring problem I've had in designing AxoPy's API is trying to keep things such that a task can be written generically and you can plug in whatever input device and pipeline you want to drive it. I think this may just be an unrealistic goal in all but the simplest use cases, and coordinating multiple input devices is definitely not a simple case. Here's an example design: user passes a dictionary with generic names for the devices:

```python
exp = Experiment(daq={'emg': TrignoEMG(...), 'accel': TrignoAccel(...)})
```

Then in your task you can connect them to separate callbacks:

```python
def prepare_daq(self, daqstream):
    daqstream['emg'].updated.connect(self.update_emg)
    daqstream['accel'].updated.connect(self.update_accel)
```

It's then up to your task implementation to blend the inputs. This could be done by creating a pipeline with a custom block with a method to take in and cache the IMU inputs. When the main processing loop runs, that block could extrapolate the most recent IMU input so you can combine the EMG and IMU (with a KF, linear extrapolation, whatever). You could actually directly connect the IMU transmitter to that block's extra method to avoid the extra task callback. This is my understanding of how people actually use the Delsys Trigno EMG+IMU for realtime control, for example, but I've never really dug into the methods people use. I would say requiring the update rates of the devices to be a simple integer factor of one another is too limiting.

This approach gives a lot of freedom over how you want to connect up and use multiple input devices, but the tasks then require knowledge of those names. This is already somewhat of a problem I've had implementing experiments with sharing pipelines between tasks and needing to access specific blocks (i.e. the task needs to know some of the "global" information about the experiment). This is maybe a bigger discussion point, but for the sake of this issue, does the approach passing a dictionary of `daq`s seem reasonable?

I should also note that even though `Experiment` only supports one `DaqStream`, there's nothing to stop you from wrapping another input device in a `DaqStream` and connecting it to a separate callback like above. This is similar to the way I just kind of left out an "output_stream" interface for controlling hardware devices or a 3D simulation environment or whatever (left as an exercise for the user :). Qt's default signal/slot connections are queued, so there shouldn't be concurrency issues with this approach though I haven't actually tested it.

**Comment by agamemnonc** [comment]
Many thanks for providing a potential solution.

First off, I agree that supporting only devices with sampling rates that are integer factors of one another would be too limiting (and wouldn't work for trigno EMG/IMU, for example).

Am I right in thinking that in order to support multiple streams the code in Experiment should be modified as follows?

https://github.com/axopy/axopy/blob/4ef815d2ea273d6a5a163c92b33234bfca82df83/axopy/experiment.py#L39

to something like:

```python
if isinstance(self.daq, dict):
    self.daqstream = dict()
    for daq_name, daq_ in daq.items():
        self.daqstream[daq_name] = DaqStream(daq_)
else:
    self.daqstream = DaqStream(daq)
```
I think this might be actually what you meant by your last paragraph, where you are suggesting having a "master" DaqStream passed into the Experiment and additional DaqStreams defined in the Task.

One particular point I am not really sure about is the following (please correct me if I am wrong).

Let's consider a simple task where streamers read data, their `updated` signals are connected to some processing `callbacks` which output some data (e.g. one of the data sources is upsampled/interpolated) and then the combined data are processed with some pipeline in another callback. At the end of that callback we increment the timer to move on to the next cycle of the main processing loop. The timer is used to control the trial length under the given streamer configuration (similar to the `adaptive_cursor_control` example).

Given the current API design, I understand that the cycle length in the main program loop is dependent upon the `rate` and `read_size` (i.e. `cycle length = read_size / rate`) of the DaqStream.

If there are multiple DaqStreams with different rates and read_sizes, then the cycle length will be determined by the device that has "slower" `read_size / rate`.  Will this not result in data being missed from the DaqStream with faster read operations (as the cycle time is longer than the rate the device samples data)? 

On a relevant note, I am not quite sure how and if it's possible to connect two or more signals to a single slot using AND logic (i.e., the callback is called only when both signals have been emitted)? 

If what I described above doesn't really make sense, it might be actually easier to discuss with the help of a basic example (using for instance two NoiseGenerator sources with different `rate` and `read_size` parameters). 

Sorry if I am missing something too obvious here!

**Comment by ixjlyons** [comment]
Yes the code you have there is essentially what I was thinking.

My last paragraph refers to a workaround to use multiple input devices *now* without requiring modifications to AxoPy. Example:

```python
daq1 = NoiseGenerator()
daq2 = Keyboard()
exp = Experiment(daq=daq1)
exp.run(MyTask())

class MyTask(Task):
    def prepare_daq(self, daqstream):
        self.noise_gen = daqstream
        self.noise_gen.updated.connect(self.update)

        daq2.updated.connect(self.kbd_input)

    def update(self, noise_data):
        ...

    def kbd_input(self, keys):
        ....
```

The main (minor) annoyance is that you'll have to make sure everything's disconnected and the global DAQ is stopped when the task finishes and hands control back to the experiment.

The problem you describe is, I think, a fundamental limitation and I can't think of a way around it unless the update rates are related by a small integer factor. One way to make sure you don't miss any data is to set a flag once the timer expires, disable the main DAQ callback connection, then wait for the next secondary DAQ update to finish the trial. From a real-time control standpoint, that last frame of data wasn't used and fed back to the subject, but at least you have the raw data in case you want to store it.

You can connect multiple Qt signals to a single slot, but I'm not aware of a built-in feature like what you're describing (kind of like a mux?). You could always implement your own logic in the callback to wait for both inputs, but it may get tricky to implement depending on how the update rates are related (you can get the object transmitting the event with `self.sender()` in the callback as long as the callback is a method of a class inheriting from `QObject`, but I haven't verified that using that and equality tests work correctly).

I totally agree some concrete examples would be helpful for illustrating some of this and verifying that my ideas here actually work. I will try to come up with something soon.

**Comment by ixjlyons** [comment]
So here's an example of the workaround I've mentioned.

```python
from axopy.experiment import Experiment
from axopy.task import Task
from axopy.daq import NoiseGenerator, DaqStream
from axopy.timing import Counter
from axopy.gui.canvas import Canvas, Text

daq1 = NoiseGenerator(read_size=10, rate=100)
daq2 = NoiseGenerator(read_size=30, rate=100)
daq2stream = DaqStream(daq2)


class CountTask(Task):

    def prepare_graphics(self, container):
        self.text1 = Text('0')
        self.text1.pos = -0.5, 0

        self.text2 = Text('0')
        self.text2.pos = 0.5, 0

        canvas = Canvas()
        canvas.add_item(self.text1)
        canvas.add_item(self.text2)
        container.set_widget(canvas)

    def prepare_daq(self, daqstream):
        self.counter1 = Counter(100, reset_on_timeout=False)
        self.counter2 = Counter(1000)

        self.counter1.timeout.connect(self.done)

        self.daq1stream = daqstream
        self.daq2stream = daq2stream
        self.daq1stream.updated.connect(self.update_daq1)
        self.daq2stream.updated.connect(self.update_daq2)

    def run(self):
        self.daq1stream.start()
        self.daq2stream.start()

    def update_daq1(self, data):
        self.counter1.increment()
        self.text1.qitem.setText(str(self.counter1.count))

    def update_daq2(self, data):
        self.counter2.increment()
        self.text2.qitem.setText(str(self.counter2.count))

    def done(self):
        self.daq1stream.updated.disconnect(self.update_daq1)
        self.daq2stream.updated.disconnect(self.update_daq2)

        # use wait=False so these don't hang up the final graphics update
        self.daq1stream.stop(wait=False)
        self.daq2stream.stop(wait=False)


Experiment(daq=daq1, subject='1', allow_overwrite=True).run(CountTask())
```

**Comment by agamemnonc** [comment]
Many thanks for providing this example.

Please feel free to close this issue if you 'd rather not include the changes discussed above to natively support multiple daqs. Otherwise, I am happy to submit a PR.

**Comment by agamemnonc** [comment]
Based on your example, I created a slightly more complicated one, where a master daq is used to control the main program update rate.  Data from two devices with different rates are processed and then combined. The raw and processed data are stored to disk. 

I am putting it here just in case it could be of use to anyone else, or perhaps as a potential example to include in the future (if you think it would be useful, of course)/

I should probably mention that it works using the modification discussed above --although this wouldn't be necessary if one used the trick you suggested above-- and I have also created a `DumbDaq` device, which is similar to `NoiseGenerator` except that it doesn't provide any data, i.e., it only blocks execution:

```python
class DumbDaq(object):
    """An emulated data acquisition device that doesn't generate any data.

    :class:`DumbGenerator` is meant to emulate data acquisition devices that
    block on each request for data until the data is available. See
    :meth:`read` for details.

    Parameters
    ----------
    rate : int, optional
        Sample rate in Hz. Default is 1000.
    read_size : int, optional
        Number of samples to generate per :meth:`read()` call. Default is 100.
    """

    def __init__(self, rate=1000, read_size=100):
        self.rate = rate
        self.read_size = read_size

        self.sleeper = _Sleeper(float(self.read_size/self.rate))

    def start(self):
        """Does nothing for this device. Implemented to follow device API."""
        pass

    def read(self):
        """
        Blocks execution.

        This method blocks (calls ``time.sleep()``) to emulate other data
        acquisition units which wait for the requested number of samples to be
        read. The amount of time to block is calculated such that consecutive
        calls will always return with constant frequency, assuming the calls
        occur faster than required (i.e. processing doesn't fall behind).
        """
        self.sleeper.sleep()

    def stop(self):
        """Does nothing for this device. Implemented to follow device API."""
        pass

    def reset(self):
        """Reset the device back to its initialized state."""
        self.sleeper.reset()
```

And the example:

```python
"""
Async daqs
==========

This examples demonstrates the use of multiple daq devices in an asyncrhonised
fashion. Daqs do not need to have sampling rates that are integer multiples of
one another, yet, it makes sense to set their read_sizes such that it takes
"roughly" the same time to perform a read operation from the various devices.
A DumbDaq is used as a master daq to control the main program update rate.
We might want to set the properties of the master daq to be equal to that of
the "faster" of the real daqs. This is however not strict; we could also set the
read time of the master daq to be even faster (but not slower). In that case,
we would end up with correlated samples as we sample faster than the daqs are
providing data.

The example demonstrates how to store the raw data from the multiple devices
and also how to combine (through concatenation) and store their processed data.
The various devices may provide series of different lengths but interpolation
is not required as the final step of the processing pipeline includes a
feature computation that is performed along the time axis.

If ``rate_m``, ``read_size_m`` and ``counter_m`` are the sampling rate, read
size and the counter limit, respectively, of the master daq, and ``rate`` and
``read_size`` are the rate and read size of a real daq, then the output raw
data array for this device is expected to have shape:
(n_channels, rate * int(counter_m * read_size_m / rate_m)).

The concatenated processed data array will have shape:
(n_total_channels, counter_m), where ``n_total_channels`` is the sum of the
number of channels of the various devices.
"""

import numpy as np
from scipy.signal import butter

from axopy.experiment import Experiment
from axopy.task import Task
from axopy.timing import Counter
from axopy.gui.canvas import Canvas, Text
from axopy.features import mean_absolute_value
from axopy.pipeline import Pipeline, Windower, Callable, Filter


class CountTask(Task):
    def __init__(self, rate, readsize):
        super().__init__()
        self.rate = rate
        self.readsize = readsize
        self.pipeline = {
            'daq_1': self.make_pipeline(rate=self.rate['daq_1'],
                                        readsize=self.readsize['daq_1'],
                                        winsize=250,
                                        lowpass=100,
                                        highpass=500),
            'daq_2': self.make_pipeline(rate=self.rate['daq_2'],
                                        readsize=self.readsize['daq_2'],
                                        winsize=15,
                                        lowpass=5,
                                        highpass=15)
        }
        # This is where the data from the multiple streams will be stored
        # after they have been processed.
        self.cur_data = {'daq_1': None, 'daq_2': None}

    def make_pipeline(self, rate, readsize, winsize, lowpass, highpass):
        b, a = butter(4, (lowpass/rate/2., highpass/rate/2.), 'bandpass')
        pipeline = Pipeline([
            Windower(winsize),
            # overlap = winsize - read_rate
            Filter(b, a=a, overlap=winsize-readsize),
            Callable(mean_absolute_value,
                     func_kwargs={'axis': 1, 'keepdims': True})
        ])

        return pipeline

    def prepare_design(self, design):
        block = design.add_block()
        block.add_trial()

    def prepare_graphics(self, container):
        self.text0 = Text('Master: 0')
        self.text0.pos = -0.2, 0.5

        self.text1 = Text('Daq 1: 0')
        self.text1.pos = -0.2, 0

        self.text2 = Text('Daq 2: 0')
        self.text2.pos = -0.2, -0.5

        canvas = Canvas()
        canvas.add_item(self.text0)
        canvas.add_item(self.text1)
        canvas.add_item(self.text2)
        container.set_widget(canvas)

    def prepare_daq(self, daqstream):
        # The master counter will determine trial time, i.e.
        # trial_time = counter_limit * read_daq / rate_daq
        self.counter0 = Counter(100, reset_on_timeout=False)
        self.counter1 = Counter(1000)
        self.counter2 = Counter(1000)

        self.counter0.timeout.connect(self.finish_trial)

        self.daqstream = daqstream

    def prepare_storage(self, storage):
        self.writer = storage.create_task('async_daqs')

    def run_trial(self, trial):
        trial.add_array('dev_1_raw', stack_axis=1)
        trial.add_array('dev_2_raw', stack_axis=1)
        trial.add_array('dev_12_processed', stack_axis=1)

        self.daqstream['daq_0'].start()
        self.daqstream['daq_1'].start()
        self.daqstream['daq_2'].start()

        self.daqstream['daq_0'].updated.connect(self.update_daq0)
        self.daqstream['daq_1'].updated.connect(self.update_daq1)
        self.daqstream['daq_2'].updated.connect(self.update_daq2)

    def update_daq0(self):
        # Daq 0 is the "master" Daq, i.e. an update happens when daq 0 is
        # updated. The check is used to ensure that updates start only after
        # the two streams have started providing data.
        if not any(elem is None for elem in self.cur_data.values()):
            daq1data = self.cur_data['daq_1'].copy()
            daq2data = self.cur_data['daq_2'].copy()
            data_c = np.concatenate((daq1data, daq2data), axis=0)

            self.trial.arrays['dev_12_processed'].stack(data_c)

            self.counter0.increment()
            self.text0.qitem.setText("Master: " + str(self.counter0.count))

    def update_daq1(self, data):
        self.counter1.increment()
        self.text1.qitem.setText("Daq 1: " + str(self.counter1.count))

        data_proc = self.pipeline['daq_1'].process(data)
        self.cur_data['daq_1'] = data_proc
        self.trial.arrays['dev_1_raw'].stack(data)

    def update_daq2(self, data):
        self.counter2.increment()
        self.text2.qitem.setText("Daq 2: " + str(self.counter2.count))

        data_proc = self.pipeline['daq_2'].process(data)
        self.cur_data['daq_2'] = data_proc
        self.trial.arrays['dev_2_raw'].stack(data)

    def finish_trial(self):
        self.writer.write(self.trial)

        self.daqstream['daq_0'].updated.disconnect(self.update_daq0)
        self.daqstream['daq_1'].updated.disconnect(self.update_daq1)
        self.daqstream['daq_2'].updated.disconnect(self.update_daq2)

        # use wait=False so these don't hang up the final graphics update
        self.daqstream['daq_0'].stop(wait=False)
        self.daqstream['daq_1'].stop(wait=False)
        self.daqstream['daq_2'].stop(wait=False)

        self.next_trial()


if __name__ == '__main__':
    from axopy.daq import NoiseGenerator, DumbDaq
    rate = {'daq_0': 1000, 'daq_1': 2000, 'daq_2': 42}
    readsize = {'daq_0': 100, 'daq_1': 200, 'daq_2': 4}

    daq0 = DumbDaq(rate=rate['daq_0'], read_size=readsize['daq_0'])
    daq1 = NoiseGenerator(num_channels=4, rate=rate['daq_1'],
                          read_size=readsize['daq_1'])
    daq2 = NoiseGenerator(num_channels=2, rate=rate['daq_2'],
                          read_size=readsize['daq_2'])

    exp = Experiment(daq={'daq_0': daq0, 'daq_1': daq1, 'daq_2': daq2},
                     subject='test', allow_overwrite=True)
    exp.run(CountTask(rate, readsize))
```

**Comment by ixjlyons** [comment]
Thanks! I'll look at your example a bit more but it definitely looks like it could go in the examples.

I am in favor of supporting multiple input devices properly. I think the basic dictionary approach discussed above is the way to go (pull requests welcome of course but I also don't mind implementing it, it'll just be slow). My concerns about the broken modularity of tasks is a bigger/different issue -- I originally was envisioning built-in tasks the user could user could easily change by overriding some behavior. This is just not the way it will work and reusable components will just have to be lower-level things like graphical layouts/widgets and maybe generic "trial flow" classes that could be used as mixins in task implementations. Basically, the `Task` is where the user connects everything up and specifies mostly everything about the experiment, so it doesn't make much sense to have built-in tasks aside from some very generic ones I've mentioned in other issues (e.g. processor task).

Also, seeing your `DumbDaq`, I thought of an idea to make a `FunctionGenerator` (maybe too cute of a name if it causes confusion regarding Python functions and generators), which works exactly like `NoiseGenerator` but takes a callable the user can provide to generate custom output (or none at all).

**Comment by agamemnonc** [comment]
Great. Just submitted a PR. 

I will create a new issue for the function generator and we can discuss there.
